src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -8216,28 +8230,55 @@ internal static bool IsReparsePoint(FileSystemInfo fileInfo)
             return fileInfo.Attributes.HasFlag(System.IO.FileAttributes.ReparsePoint);
         }
 
-        internal static bool IsReparsePointWithTarget(FileSystemInfo fileInfo)
+        internal static bool IsReparsePointLikeSymlink(FileSystemInfo fileInfo)
         {
-            if (!IsReparsePoint(fileInfo))
+#if UNIX
+            // Reparse point on Unix is a symlink.
+            return IsReparsePoint(fileInfo);
+#else
+            if (InternalTestHooks.OneDriveTestOn && fileInfo.Name == InternalTestHooks.OneDriveTestSymlinkName)
             {
-                return false;
+                return !InternalTestHooks.OneDriveTestRecurseOn;
             }
-#if !UNIX
-            // It is a reparse point and we should check some reparse point tags.
-            var data = new WIN32_FIND_DATA();
-            using (var handle = FindFirstFileEx(fileInfo.FullName, FINDEX_INFO_LEVELS.FindExInfoBasic, ref data, FINDEX_SEARCH_OPS.FindExSearchNameMatch, IntPtr.Zero, 0))
+
+            WIN32_FIND_DATA data = default;
+            string fullPath = Path.TrimEndingDirectorySeparator(fileInfo.FullName);
+            if (fullPath.Length > MAX_PATH)
+            {
+                fullPath = PathUtils.EnsureExtendedPrefix(fullPath);
+            }
+
+            using (var handle = FindFirstFileEx(fullPath, FINDEX_INFO_LEVELS.FindExInfoBasic, ref data, FINDEX_SEARCH_OPS.FindExSearchNameMatch, IntPtr.Zero, 0))
             {
+                if (handle.IsInvalid)
+                {
+                    // Our handle could be invalidated by something else touching the filesystem,
+                    // so ensure we deal with that possibility here
+                    int lastError = Marshal.GetLastWin32Error();
+                    throw new Win32Exception(lastError);
+                }
+
+                // We already have the file attribute information from our Win32 call,
+                // so no need to take the expense of the FileInfo.FileAttributes call
+                const int FILE_ATTRIBUTE_REPARSE_POINT = 0x0400;
+                if ((data.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) == 0)
```

Comment:
By doing this check before `FindFirstFileEx` you save a call from happening on the dir which can add up when dealing with SMB paths or lots of files in a recursive `gci`. We should add the `IsReparsePoint(fileInfo)` and return if `false` check before the `FindFirstFileEx` call to avoid all this.

Reply:
In the scenario we call IsReparsePointLikeSymlink only if checkReparsePoint is true - that is de-facto we already have called  IsReparsePoint.

Reply:
Is that true? I cannot seem to find that logic and can see multiple ways where this is not the case. It also means we do these unnecessary checks anyway for Unix and here as well.

Even if it was actually the case it's still good to optimize the code in case something in the future calls `IsReparsePointLikeSymlink` and doesn't check beforehand.

Reply:
> Is that true? I cannot seem to find that logic

See `checkReparsePoint` in the PR.

> can see multiple ways where this is not the case.

There are only 4 places where we call IsReparsePointLikeSymlink and I don't see where we need to worry about regression.

> Even if it was actually the case it's still good to optimize the code in case something in the future calls IsReparsePointLikeSymlink and doesn't check beforehand.

The perf optimization removes extra p/invokes calls - no need do p/invoke with IsReparsePoint() if we need only check target existence.

As for future, it is internal code and we free to change it as needed. 

But most likely this code will be removed altogether soon (I expect it will be in .Net 7.0), since there is a discussion going on about adding these features to .Net (symlink and reparsepoint tags API, and appropriate enhancements to path enumeration API).

Reply:
> There are only 4 places where we call IsReparsePointLikeSymlink and I don't see where we need to worry about regression.

It's not a regression but an improvement.

* https://github.com/PowerShell/PowerShell/blob/35ebe2c9ac36810b369decf807686371815c31f5/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L2067 - No check for reparse point `IsReparsePointLikeSymlink` is called regardless
* https://github.com/PowerShell/PowerShell/blob/35ebe2c9ac36810b369decf807686371815c31f5/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L3107 - Removal code, `IsReparsePointLikeSymlink` is called without a reparse point check

So instead of adding a check in `Dir` you are better off doing it just once in `IsReparsePointLikeSymlink` so it all gets it by default.

> The perf optimization removes extra p/invokes calls - no need do p/invoke with IsReparsePoint() if we need only check target existence.

Only for the `Dir` function, the other 2 I linked to doesn't have this optimisation. My point is put the `ReparsePoint` check in `IsReparsePointLikeSymlink` instead so all 3 invocations can take advantage of this optimisation and you can remove the extra code in `Dir`.

Reply:
> Only for the Dir function, the other 2 I linked to doesn't have this optimisation. 

No :-) Exactly the opposite.

We have 2 API where we do (expensive) p/invokes - (1) in File.Attributes properties (follow calls use a cache) - used in IsReparsePoint(), (2) FindFirstFileEx, used in IsReparsePointLikeSymlink
If we need only FindFirstFileEx we shouldn't do first p/invoke since FindFirstFileEx returns all we need.

Reply:
Hmm I think I see what you are saying now, `File/Directory.Attributes` will call `FindFirstFileEx` internally but only on demand. not when the object is constructed. So since `Dir` already needed the `Attributes` for another call we already had that to skip the subsequent call but the others haven't had that optimisation yet.